### PR TITLE
Valid underscores in hex/octal/binary literal docs

### PR DESCRIPTION
**Pull request by syvb**
Currently hex/octal/binary literals with computed values are displayed like `0_xff_fff_fffu32`, which is invalid since underscores can't be in the middle of integer prefixes. This properly formats prefixed integers.

This causes  [`std::u32::MAX`](https://doc.rust-lang.org/std/u32/constant.MAX.html) to be displayed as
```rust
pub const MAX: u32 = u32::MAX; // 0_xff_fff_fffu32
```

This PR changes it to be displayed as:
```rust
pub const MAX: u32 = u32::MAX; // 0xffff_ffffu32
```

### CODE REVIEW

File: src/librustdoc/clean/utils/tests.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    assert_eq!(format_integer_with_underscore_sep("-0o123"), "-0o123");
+    assert_eq!(format_integer_with_underscore_sep("-0o123456"), "-0o12_3456");
+}
+
+#[test]
+fn int_format_octal() {
+    assert_eq!(format_integer_with_underscore_sep("0b101"), "0b101");
+    assert_eq!(format_integer_with_underscore_sep("0b101101011"), "0b1_0110_1011");
+    assert_eq!(format_integer_with_underscore_sep("-0b101"), "-0b101");
+    assert_eq!(format_integer_with_underscore_sep("-0b101101011"), "-0b1_0110_1011");
```
Comment (pickfire) [inline src/librustdoc/clean/utils/tests.rs:32]: Can we add a test on the boundary like `-0b0110_1011`?

**Review by pickfire — APPROVED** [review]
Looks good to me. But I wonder why the `tests` couldn't be kept in the same file?

**Comment by Smittyvb** [comment]
@pickfire All of the other Rustdoc tests are also in their own `tests` module, so I added these tests using the same convention.

File: src/librustdoc/clean/utils.rs [hunk 2]
```diff
@@ -331,11 +334,21 @@ crate fn print_evaluated_const(tcx: TyCtxt<'_>, def_id: DefId) -> Option<String>
...
+    let mut num_start_index = if num_chars.get(0) == Some(&'-') { 1 } else { 0 };
+    let chunk_size = if matches!(
+        (num_chars.get(num_start_index), num_chars.get(num_start_index + 1)),
+        (Some(&'0'), Some(&'b' | &'o' | &'x'))
+    ) {
+        num_start_index += 2;
+        4
+    } else {
+        3
+    };
```
Comment (agurney) [inline src/librustdoc/clean/utils.rs:347]: I think it might be nice if this matched what clippy suggests for octal literals in https://github.com/rust-lang/rust/blob/master/src/tools/clippy/clippy_utils/src/numeric_literal.rs#L18, which is groups of 3 rather than 4. I don't know why 3, but maybe it's because of Unix permissions?

File: src/librustdoc/clean/utils/tests.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+use super::*;
+
+#[test]
+fn int_format_decimal() {
+    assert_eq!(format_integer_with_underscore_sep("12345678"), "12_345_678");
+    assert_eq!(format_integer_with_underscore_sep("123"), "123");
+    assert_eq!(format_integer_with_underscore_sep("123459"), "123_459");
+    assert_eq!(format_integer_with_underscore_sep("-12345678"), "-12_345_678");
+    assert_eq!(format_integer_with_underscore_sep("-123"), "-123");
+    assert_eq!(format_integer_with_underscore_sep("-123459"), "-123_459");
```
Comment (jyn514) [inline src/librustdoc/clean/utils/tests.rs:10]: These unit tests are amazing, thank you :heart:

File: src/librustdoc/clean/utils.rs [hunk 4]
```diff
@@ -331,11 +334,27 @@ crate fn print_evaluated_const(tcx: TyCtxt<'_>, def_id: DefId) -> Option<String>
...
 fn format_integer_with_underscore_sep(num: &str) -> String {
     let num_chars: Vec<_> = num.chars().collect();
-    let num_start_index = if num_chars.get(0) == Some(&'-') { 1 } else { 0 };
+    let mut num_start_index = if num_chars.get(0) == Some(&'-') { 1 } else { 0 };
+    let chunk_size = match (num_chars.get(num_start_index), num_chars.get(num_start_index + 1)) {
+        (Some('0'), Some('b' | 'x')) => {
+            num_start_index += 2;
+            4
+        }
+        (Some('0'), Some('o')) => {
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:343]:
```suggestion
    let chunk_size = match num.as_bytes() {
        [b'0', b'b', ..] => {
            num_start_index += 2;
            4
        }
        [b'0', b'o', ..] => {
```

File: src/librustdoc/clean/utils.rs [hunk 5]
```diff
@@ -331,11 +334,27 @@ crate fn print_evaluated_const(tcx: TyCtxt<'_>, def_id: DefId) -> Option<String>
 
 fn format_integer_with_underscore_sep(num: &str) -> String {
     let num_chars: Vec<_> = num.chars().collect();
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:336]: A good follow-up would be to remove `num_chars` altogether - `num` is always ascii so you should be able to just use bytes directly. Doesn't need to be in this PR though :)

**Review by jyn514 — APPROVED** [review]
> @pickfire All of the other Rustdoc tests are also in their own `tests` module, so I added these tests using the same convention.

Yup, and this also avoids recompiling the main rustdoc crate when you only change a test.

This looks good to me, just a couple nits on the implementation :)

File: src/librustdoc/clean/utils.rs [hunk 6]
```diff
@@ -335,12 +335,12 @@ crate fn print_evaluated_const(tcx: TyCtxt<'_>, def_id: DefId) -> Option<String>
...
     let mut num_start_index = if num_chars.get(0) == Some(&'-') { 1 } else { 0 };
-    let chunk_size = match (num_chars.get(num_start_index), num_chars.get(num_start_index + 1)) {
-        (Some('0'), Some('b' | 'x')) => {
+    let chunk_size = match num.as_bytes() {
+        [b'0', b'b', ..] => {
```
Comment (jyn514) [inline src/librustdoc/clean/utils.rs]: Oh oops, I suggested a change in behavior,  sorry.
```suggestion
        [b'0', b'b' | b'x', ..] => {
```

Comment (Smittyvb) [reply src/librustdoc/clean/utils.rs]: ced3c5ad26857e2c426306d821641e91dc8e860d

**Comment by jyn514** [comment]
Do you mind squashing the commits into one? There are instructions at https://rustc-dev-guide.rust-lang.org/git.html#advanced-rebasing, or if it's a pain bors can do it automatically (it will just show the PR as closed instead of merged because github doesn't recognize it).

**Comment by jyn514** [comment]
@bors r+

Thanks for working on this!
